charm/ksql/reactive/ksql.py

```diff
@@ -0,0 +1,234 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import socket
+import tempfile
+
+from pathlib import Path
+from OpenSSL import crypto
+from subprocess import check_call
+
+from charms.reactive import set_state, remove_state, when, when_not, hook, when_file_changed
+from charmhelpers.core import hookenv
+from charms.reactive.helpers import data_changed
+from charmhelpers.core.hookenv import log
+
+from charms.layer.ksql import KSQL_SNAP, KSQL_SNAP_COMMON, KSQL_PORT, Ksql
+from charms.layer import tls_client
+
+@when_not('ksql.available')
+def install():
+    install_snap()
+
+@hook('upgrade-charm')
+def upgrade():
+    upgrade_snap()
+
+@hook('stop')
+def uninstall():
+    try:
+        check_call(['snap', 'remove', KSQL_SNAP])
+    except subprocess.CalledProcessError as e:
+        hookenv.log("failed to remove snap: {}".format(e))
+
+def install_snap():
+    check_call(['snap', 'install', '--edge', KSQL_SNAP])
```

Comment:
This is OK for now, but eventually we'll promote this snap to stable. Will need to address this in a followup, perhaps a config option?

Reply:
Included a channel option in config.yml